### PR TITLE
Vendor pack CLI code to build with Buildpacks

### PR DESCRIPTION
**Pull request by dgageot**

### CODE REVIEW

File: pkg/skaffold/build/buildpacks/lifecycle.go [hunk 1]
```diff
@@ -18,19 +18,21 @@ package buildpacks
...
 
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/build/misc"
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/docker"
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/latest"
-	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/util"
+)
+
+// For testing
+var (
+	runPackFunc = runPack
```
Comment (briandealwis) [inline pkg/skaffold/build/buildpacks/lifecycle.go]: Optional: rename runPack -> runPackBuild just to be very obvious?

File: pkg/skaffold/build/buildpacks/logger.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+import (
+	"io"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/buildpacks/pack/logging"
+)
+
+type logger struct {
```
Comment (briandealwis) [inline pkg/skaffold/build/buildpacks/logger.go]: Would be worth a comment here to say this logger exists to meet the requirements of the pack logger.

**Comment by dgageot** [comment]
@briandealwis you should be able to LGTM since all the changes are buildpacks related now. The go-licences part was merged.

File: pkg/skaffold/build/buildpacks/lifecycle.go [hunk 3]
```diff
@@ -48,17 +50,25 @@ func (b *Builder) build(ctx context.Context, out io.Writer, a *latest.Artifact,
 
 	builderImage := artifact.Builder
 	logrus.Debugln("Builder image", builderImage)
-	if err := b.pull(ctx, out, builderImage, artifact.ForcePull); err != nil {
-		return "", err
+	if !artifact.ForcePull {
+		if err := b.pull(ctx, out, builderImage); err != nil {
+			return "", err
```
Comment (briandealwis) [inline pkg/skaffold/build/buildpacks/lifecycle.go]: This is confusing: if we aren't force-pulling then we pull the image?  Doesn't pack do this anyways?  Or this needs a comment to explain why.

File: pkg/skaffold/build/buildpacks/lifecycle.go [hunk 4]
```diff
@@ -48,17 +50,25 @@ func (b *Builder) build(ctx context.Context, out io.Writer, a *latest.Artifact,
...
-		return "", err
-	}
-	logrus.Debugln("Run image", runImage)
-	if err := b.pull(ctx, out, runImage, artifact.ForcePull); err != nil {
-		return "", err
+	runImage := artifact.RunImage
+	if !artifact.ForcePull {
+		// If ForcePull is true, we let pack find and pull the run image
+		var err error
+		runImage, err = b.findRunImage(ctx, artifact, builderImage)
```
Comment (briandealwis) [inline pkg/skaffold/build/buildpacks/lifecycle.go]: Doesn't pack already do this?

Comment (dgageot) [reply pkg/skaffold/build/buildpacks/lifecycle.go]: pack is either pulling always or pulling never. This is trying to `pullIfNotPresent`. I'll update the code with a comment.

File: pkg/skaffold/build/buildpacks/lifecycle.go (hunk 3 above)
Comment (dgageot) [reply pkg/skaffold/build/buildpacks/lifecycle.go]: pack is either pulling always or pulling never. This is trying to pullIfNotPresent. I'll update the code with a comment.

File: pkg/skaffold/build/buildpacks/lifecycle.go [hunk 5]
```diff
@@ -48,17 +50,30 @@ func (b *Builder) build(ctx context.Context, out io.Writer, a *latest.Artifact,
 
 	builderImage := artifact.Builder
 	logrus.Debugln("Builder image", builderImage)
-	if err := b.pull(ctx, out, builderImage, artifact.ForcePull); err != nil {
-		return "", err
+	// If ForcePull is false: we pull the image only if it's not there already.
+	// If ForcePull is true: we will let `pack` always pull.
+	// Ideally, we add a `--pullIdNotPresent` option to upstream `pack`.
```
Comment (briandealwis) [inline pkg/skaffold/build/buildpacks/lifecycle.go]:
```suggestion
	// Ideally, we add a `--pullIfNotPresent` option to upstream `pack`.
```

File: pkg/skaffold/build/buildpacks/lifecycle.go [hunk 6]
```diff
@@ -48,17 +50,30 @@ func (b *Builder) build(ctx context.Context, out io.Writer, a *latest.Artifact,
...
-		return "", err
-	}
-	logrus.Debugln("Run image", runImage)
-	if err := b.pull(ctx, out, runImage, artifact.ForcePull); err != nil {
-		return "", err
+	runImage := artifact.RunImage
+	if !artifact.ForcePull {
+		// If ForcePull is false: we pull the image only if it's not there already.
+		// If ForcePull is true: we will let `pack` always pull.
+		// Ideally, we add a `--pullIdNotPresent` option to upstream `pack`.
```
Comment (briandealwis) [inline pkg/skaffold/build/buildpacks/lifecycle.go]:
```suggestion
		// Ideally, we add a `--pullIfNotPresent` option to upstream `pack`.
```

**Pull request merged**